labelbox/pagination.py

```diff
@@ -25,45 +40,101 @@ def __init__(self, client, query, params, dereferencing, obj_class):
+            cursor_path: If not None, this is used to find the cursor
+            experimental: Used to call experimental endpoints
         """
         self.client = client
         self.query = query
         self.params = params
         self.dereferencing = dereferencing
         self.obj_class = obj_class
+        self.cursor_path = cursor_path
+        self.beta = beta
 
-        self._fetched_pages = 0
         self._fetched_all = False
-        self._data = []
+        self._data: List[Dict[str, Any]] = []
         self._data_ind = 0
 
+        self.paginator = _CursorPagination(
+            client, cursor_path) if cursor_path else _OffsetPagination(client)
+
     def __iter__(self):
         self._data_ind = 0
         return self
 
+    def get_page_data(self, results):
+        for deref in self.dereferencing:
+            results = results[deref]
+        return [self.obj_class(self.client, result) for result in results]
+
     def __next__(self):
         if len(self._data) <= self._data_ind:
             if self._fetched_all:
                 raise StopIteration()
 
-            query = self.query % (self._fetched_pages * _PAGE_SIZE, _PAGE_SIZE)
-            self._fetched_pages += 1
-
-            results = self.client.execute(query, self.params)
-            for deref in self.dereferencing:
-                results = results[deref]
-
-            page_data = [
-                self.obj_class(self.client, result) for result in results
-            ]
+            results = self.paginator.fetch_results(self.query, self.params,
+                                                   self.beta)
+            page_data = self.get_page_data(results)
             self._data.extend(page_data)
+            n_items = len(page_data)
 
-            if len(page_data) < _PAGE_SIZE:
-                self._fetched_all = True
-
-            if len(page_data) == 0:
+            if n_items == 0:
                 raise StopIteration()
 
+            self._fetched_all = self.paginator.fetched_all(n_items, results)
+
         rval = self._data[self._data_ind]
         self._data_ind += 1
         return rval
+
+
+class _Pagination(ABC):
+
+    @abstractmethod
+    def fetched_all(self, n_items: int, results: List[Dict[str, Any]]) -> bool:
+        ...
+
+    @abstractmethod
+    def fetch_results(self, query: str, params: Dict[str, Any],
+                      beta: bool) -> Dict[str, Any]:
+        ...
+
+
+class _CursorPagination(_Pagination):
+
+    def __init__(self, client: "Client", cursor_path: Dict[str, Any]):
+        self.client = client
+        self.cursor_path = cursor_path
+        self.next_cursor: Optional[str] = None
+
+    def get_next_cursor(self, results) -> Optional[str]:
+        for path in self.cursor_path:
+            results = results[path]
+        return results
+
+    def fetched_all(self, n_items: int, results: List[Dict[str, Any]]) -> bool:
+        self.next_cursor = self.get_next_cursor(results)
```

Comment:
Imho it's a bit confusing that `fetched_all` updates the cursor. Judging by the name, it should be an idempotent operation. Could we add a new method to the interface and move updating cursor there?

Reply:
Sounds good

Reply:
Fixed.